consensus/proto_array/src/proto_array.rs

```diff
@@ -300,6 +328,204 @@ impl ProtoArray {
         }
     }
 
+    /// Potentially updates blocks to have an invalid payload and therefore be ineligible for the
+    /// head.
+    ///
+    /// The `head_block_root` is the block *root* of the latest invalid block. The
+    /// `latest_valid_hash` is a payload *hash* and should be:
+    ///
+    /// - `Some(hash)` if the block with that payload *hash* is known to be valid and is an
+    ///     ancestor of `head_block_root`.
+    /// - `None` if the latest valid ancestor of `head_block_root` is unknown.
+    ///
+    /// ## Details
+    ///
+    /// If `head_block_root` is not known to fork choice, an error is returned.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is either not known to fork choice
+    /// (perhaps it's junk or pre-finalization), then only the `head_block_root` block will be
+    /// invalidated (no ancestors). No error will be returned in this case.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is a known ancestor of
+    /// `head_block_root`, then all blocks between `head_block_root` and `latest_valid_hash` will
+    /// be invalidated. Additionally, all blocks that descend from a newly-invalidated block will
+    /// also be invalidated.
```

Comment:
I think this comment needs to reworked now that the code has changed.

Specifically

> The `head_block_root` is the block *root* of the latest invalid block.

As discussed [here](https://github.com/sigp/lighthouse/pull/2837#discussion_r813545926) the block root we pass isn't necessarily the block root of an invalid block. It could be the parent of an invalid block.

Here's my attempt at an updated description:

```
Invalidate the relevant ancestors and descendants of a block with an invalid execution payload.

The `head_block_root` should be the beacon block root of the block with the invalid execution payload, _or_ its parent.

The `latest_valid_hash` should be the hash of most recent *valid* execution payload contained in an ancestor block of `head_block_root`.

This function will invalidate:

* The block matching `head_block_root` _unless_ that block has a payload matching `latest_valid_hash`.
* All ancestors of `head_block_root` back to the block with payload matching `latest_valid_hash` (endpoint > exclusive). In the case where the `head_block_root` is the parent of the invalid block and itself matches `latest_valid_hash`, no ancestors will be invalidated.
* All descendants of `latest_valid_hash` if supplied and consistent with `head_block_root`, or else all descendants of `head_block_root`.
```

Reply:
Good point! I've adapted a slightly modified version of your suggestion here: https://github.com/sigp/lighthouse/pull/2837/commits/6dee54c860eefe1bbc5a70d7ebc5353b1dc8f288. Thanks!